internal/ascprails/runtime.go

```
@@ -0,0 +1,375 @@
+	}
+	groups := make(map[string]*anchorGroup)
+	seenProviders := make(map[string]struct{})
+	for _, observation := range result.Observations {
+		if observation.ChainID != chainID || !identifierPattern.MatchString(observation.Provider) ||
+			observation.AnchorNumber == 0 || observation.HeadNumber < observation.AnchorNumber ||
+			!nonZeroHash(observation.AnchorHash) || !nonZeroHash(observation.HeadHash) ||
+			observation.AnchorTime.Unix() <= 0 || observation.HeadTime.Before(observation.AnchorTime) || observation.ObservedAt.IsZero() {
+			continue
+		}
+		if _, duplicate := seenProviders[observation.Provider]; duplicate {
+			return ChainObservation{}, errors.New("Base observer returned a duplicate provider")
```

golangci-lint (2.12.2)

[error] ST1005: error strings should not be capitalized

(staticcheck)